source/common/grpc/async_client_impl.cc

```diff
@@ -321,6 +337,23 @@ void AsyncStreamImpl::cleanup() {
     dispatcher_->deferredDelete(
         LinkedObject<AsyncStreamImpl>::removeFromList(parent_.active_streams_));
   }
+  remote_close_timer_ = nullptr;
+}
+
+void AsyncStreamImpl::waitForRemoteCloseAndDelete() {
+  if (!waiting_to_delete_on_remote_close_) {
+    waiting_to_delete_on_remote_close_ = true;
+
+    if (options_.sidestream_watermark_callbacks != nullptr) {
+      stream_->removeWatermarkCallbacks();
+      options_.sidestream_watermark_callbacks = nullptr;
+    }
+    remote_close_timer_ = dispatcher_->createTimer([this] {
+      waiting_to_delete_on_remote_close_ = false;
+      cleanup();
```

Comment:
if trailers come back before the timer fires, cleanup() is already called on line 270.

Reply:
But this time it will be called with the `waiting_to_delete_on_remote_close_ == false` and will call reset.

Reply:
ahh, I see waiting_to_delete_on_remote_close_ is guarding http_reset in cleanup() now.

OTOH, onTrailers --> cleanup() resets remote_close_timer_, but not disable the timer before that, does that mean the deleted timer will never get triggered, or is the timer triggered on deletion?

Reply:
Timer is not triggered on deletion.

Reply:
You are correct, the timer will not rigger and the stream will not be reset iff the client already half-closed its stream. This is the expected behavior and verified in the `GrpcClientIntegrationTest.BasicStreamDeleteOnRemoteClose` test.

If the server sends trailers before the timer, the timer is disarmed and not going to fire, the cleanup() is called from the onTrailers and the AsyncStreamImpl object is deleted. However the router filter will reset the stream when it receives the end_stream from the server before the client has half_closed. This is router's (and also Google gRPC client's) behavior. This is validated in the BasicStreamWithServerReset test.

Reply:
Indeed when trailers come back, the underneath http connection/stream impl will clear the stream itself. 
do we need "waiting_to_delete_on_remote_close_ = false;" in the timer callback?